ncm-filesystems/src/main/perl/filesystems.pm

```diff
@@ -1,116 +1,143 @@
+    $fsh{$_->{mountpoint}} = $_ foreach @$fsl;
+    return %fsh;
 }
 
-# Frees space on the system. It removes filesystems that are not
-# present in the profile anymore, and are *not* present in the
-# protected hash either. Returns 0 on success, -1 on error.
-sub free_space
+#  Removes filesystem and deletes stuff that is not present in the %mounts 
+#  argument from fstab.
+sub delete_outdated
 {
-	my ($self, $cfg, $protected, @fs) = @_;
-
-	my %ph = %$protected;
-	my %fsh = fshash (\@fs);
-
-	my $fl = output ("grep", "^[[:space:]]*#", "/etc/fstab", "-v");
-
-	$self->info ("Checking filesystems not defined in the profile");
-	my @fstab = split ("\n+", $fl);
-	foreach my $l (@fstab) {
-		$self->debug (5,"Fstab line: $l");
-		$l =~ m{^\S+\s+(\S+)\s} or next;
-		my $mount = $1;
-		next if(exists $fsh{$mount} || exists $ph{$mount}
-			    || exists($ph{abs_path($mount)}));
-		$self->verbose("Removing filesystem $mount");
-		my $f = NCM::Filesystem->new_from_fstab ($l, log => $self);
-		$f->remove_if_needed==0 or return -1;
-
-	}
-	return 0;
+    my ($self, $fstab, $mountsr, $config, $remove) = @_;
+
+    my @rm;
+    my %mounts = %$mountsr;
+    seek($fstab, 0, SEEK_SET);
+
+    while (my $f = <$fstab>) {
+    	my $mount = $self->mount_from_entry($f) or next;
+    	if (!exists ($mounts{$mount}) && !exists($mounts{abs_path($mount)})) {
+    	    CAF::Process->new ([$UMOUNT, $mount],
+    			       log => $self)->run();
+    	    push (@rm, $f);
+            $self->verbose ("Scheduling for removal of fstab: $mount");
+            $self->verbose ("Scheduling for removal filesystem $mount") if $remove;
+    	}
+    }
+
+    foreach my $outdated (@rm) {
+        if ($remove) {
+            my $fsrm = NCM::Filesystem->new_from_fstab ($outdated, $config, log => $self);
+            $self->info("Removing filesystem for $fsrm->{mountpoint}");
+            $fsrm->remove_if_needed==0 or return 0;
+        }
+        $self->info("Removing line $outdated");
+        $fstab->replace_lines (qr{$outdated}, qr{^$}, "");
+
+    }
+    return 1;
 }
 
+
+sub create_blockdevices
+{
+    my ($self, $config, $fs) = @_;
+    # Partitions must be created first
+    my $el = $config->getElement ("/system/blockdevices/partitions");
+    my @part = ();
+    $self->info ("Checking whether partitions need to be created");
+
+    while ($el && $el->hasNextElement) {
+        my $el2 = $el->getNextElement;
+        push (@part, NCM::Partition->new ($el2->getPath->toString, $config, log => $self));
+    }
+    foreach my $partition (sort partition_compare @part) {
+        if ($partition->create != 0) {
+            $self->error("Couldn't create partition: " . $partition->devpath);
+            return 0;
+        }
+    }
+    foreach my $filesystem (@$fs) {
+        if ($filesystem->create_if_needed != 0) {
+            $self->error("Failed to create filesystem:  $filesystem->{mountpoint}");
+            return 0;
+        }
+    }
+    return 1;
+};
+
+
 sub Configure
```

Comment:
how different is this code from the fstab one? if we go for aliases, i propose to make the `Configure` code equal, by adding noop subs to the fstab code or or using `$self->can()` to tets if th emethod exists.